### PR TITLE
Add volume buttons to gestures

### PR DESCRIPTION
**Pull request by tobrie**
## Purpose / Description
This commit adds the volume buttons to the gestures as discussed in #3021. Since this is unexpected behavior for most users, it's turned off by default. I realize buttons are not gestures, but it seemed like the most sensible place to put this, mainly because you can configure them for any action.

Perhaps it makes sense to add more gestures using the volume buttons and maybe even the wakup/standby button. Things like double press and press and hold could allow for more actions to be used with hardware buttons.

An interesting continuation of this would also be to make the Reviewer work with the screen turned off and TTS as output. Useful on the go and for blind people. Maybe I'll look into that.

## Approach
Simply added volume buttons to existing gestures. `AbstractFlashcardViewer.dispatchKeyEvent` then executes the assigned actions.

## How Has This Been Tested?
1. Adjust the settings (Settings > Gestures > Volume Up/Down)
2. Open deck
3. press volume button

(Only tested on an emulator)

## Checklist
- [x] You have not changed whitespace unnecessarily (it makes diffs hard to read)
- [x] You have a descriptive commit message with a short title (first line, max 50 chars).
- [x] Your code follows the style of the project (e.g. never omit braces in `if` statements) 
- [x] You have commented your code, particularly in hard-to-understand areas
- [x] You have performed a self-review of your own code

### CODE REVIEW

**Comment by mikehardy** [comment]
Sounds good in general, just a note of "caution" on trying to get AnkiDroid (or any Android app) to work with the screen off. For it to be possible I think you must have a "foreground service" connected to the work you are doing and what works out to be sort of a very fancy notification, or the system shuts you down, and we have no infrastructure for that. Nothing in Android is actually that hard to throw together to show it is possible, but I'm sure there will be little quirks to iron out afterwards. I'm open to that as it would be the same infrastructure people would want for a lock screen card viewer experience which I otherwise have no interest in coding myself but is a popular request :)

File: AnkiDroid/src/main/java/com/ichi2/anki/AbstractFlashcardViewer.java [hunk 1]
```diff
@@ -1338,6 +1340,28 @@ protected void answerCard(int ease) {
...
+            int gesture = (keyCode == KeyEvent.KEYCODE_VOLUME_UP) ? mGestureVolumeUp : COMMAND_NOTHING;
+            gesture = (keyCode == KeyEvent.KEYCODE_VOLUME_DOWN) ? mGestureVolumeDown : gesture;
+
+            // Execute command, but only consume event if action took place. We want the volume buttons to work normally otherwise.
+            if (gesture != COMMAND_NOTHING) {
+                executeCommand(gesture);
+                return true;
+            }
+
+            return false;
```
Comment (mikehardy) [inline AnkiDroid/src/main/java/com/ichi2/anki/AbstractFlashcardViewer.java:1358]: This is a bit subtle for me.

First can you do a test for ACTION_DOWN first and just return fast, that will clear that part out of the logic

Second, and you do a switch on keyCode for up, down, and default, that should cleanly separate the three cases of mGestureVolumeUp / mGestureVolumeDown / COMMAND_NOTHING without having ternary operators that might overwrite a value with itself (a concept that is a bit hard to follow)

Comment (tobrie) [reply AnkiDroid/src/main/java/com/ichi2/anki/AbstractFlashcardViewer.java]: Okay, I've changed it to be a bit more explicit. An early return would have left me with two calls to the `super` method, hence the large if. The `switch` doesn't have a default case, because gesture is initialized before and thus only declared once. Feel free to change it though, only my opinion.

**Comment by Arthur-Milchior** [comment]
Thanks for creating this PR. I love the principle. I would not have thought about it, but pressing sounds up/down seems to be cool buttons for actions

**Comment by tobrie** [comment]
By the way: Should I rebase if we merge, in order to keep the commit history pretty?

**Comment by mikehardy** [comment]
Flagging for @david-allison-1 to take a look since I know he's very interested in reviewer commands from perspective of button mapping etc

File: AnkiDroid/src/main/java/com/ichi2/anki/AbstractFlashcardViewer.java [hunk 2]
```diff
@@ -1677,6 +1705,8 @@ protected SharedPreferences restorePreferences() {
             mGestureTapTop = Integer.parseInt(preferences.getString("gestureTapTop", "12"));
             mGestureTapBottom = Integer.parseInt(preferences.getString("gestureTapBottom", "2"));
             mGestureLongclick = Integer.parseInt(preferences.getString("gestureLongclick", "11"));
+            mGestureVolumeUp = Integer.parseInt(preferences.getString("gestureVolumeUp", "11"));
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/anki/AbstractFlashcardViewer.java:1708]: We should refactor this to take a `ViewerCommandDef` (not necessarily in this PR, but the meaning of the strings isn't clear)
```suggestion
            mGestureVolumeUp = Integer.parseInt(preferences.getString("gestureVolumeUp", "0"));
```

File: AnkiDroid/src/main/java/com/ichi2/anki/AbstractFlashcardViewer.java [hunk 3]
```diff
@@ -1677,6 +1705,8 @@ protected SharedPreferences restorePreferences() {
...
+            mGestureVolumeDown = Integer.parseInt(preferences.getString("gestureVolumeDown", "11"));
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/anki/AbstractFlashcardViewer.java:1709]:
```suggestion
            mGestureVolumeDown = Integer.parseInt(preferences.getString("gestureVolumeDown", "0"));
```

File: AnkiDroid/src/main/java/com/ichi2/anki/AbstractFlashcardViewer.java [hunk 4]
```diff
@@ -1338,6 +1340,32 @@ protected void answerCard(int ease) {
     }
 
 
+    @Override
+    public boolean dispatchKeyEvent(KeyEvent event) {
+        if (event.getAction() == KeyEvent.ACTION_DOWN) {
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/anki/AbstractFlashcardViewer.java:1345]: nit: invert if to reduce nesting

Comment (tobrie) [reply AnkiDroid/src/main/java/com/ichi2/anki/AbstractFlashcardViewer.java:1345]: That would require 2 returns with calls to `super`. If you prefer I can still change it.

File: AnkiDroid/src/main/java/com/ichi2/anki/AbstractFlashcardViewer.java (hunk 3 above)
Comment (tobrie) [reply AnkiDroid/src/main/java/com/ichi2/anki/AbstractFlashcardViewer.java]: makes sense

File: AnkiDroid/src/main/java/com/ichi2/anki/AbstractFlashcardViewer.java (hunk 4 above)
Comment (david-allison-1) [reply AnkiDroid/src/main/java/com/ichi2/anki/AbstractFlashcardViewer.java:1345]: Implementer's choice

**Comment by tobrie** [comment]
Yes that is the intention.

**Comment by mikehardy** [comment]
Hello! with the 2.14 release we started accepting donations with the stated goal of accepting money from users (the users' way of saying "Thanks!") to support maintenance and development of AnkiDroid.

We are looking through PRs merged in releases going back to release 2.10 (where we have good record keeping of what's in releases) to let our contributors know.

If you'd like to submit an expense related to your work here, to share in a portion of our hopefully-sustainable monthly disbursement, please read here and do so in the next couple of days:

https://github.com/ankidroid/Anki-Android/wiki/OpenCollective-Payment-Process

This is our first time handling donations and attempting to compensate anyone with these processes, so please be flexible and patient with us

Thank you for your contribution to AnkiDroid!
